src/rttable.c

```diff
@@ -157,18 +157,24 @@ static void sendJoinLeaveUpstream(struct RouteTable* route, int join) {
                 my_log(LOG_ERR, 0 ,"FATAL: Unable to get Upstream IF.");
             }
 
-            // Check if there is a white list for the upstram VIF
+            // Check if there is a black- or whitelist for the upstram VIF
             if (upstrIf->allowedgroups != NULL) {
-              uint32_t           group = route->group;
-                struct SubnetList* sn;
+                bool                 block_list = false;
+                struct SubnetList   *match = NULL;
+                struct SubnetList   *sn;
+                uint32_t             group = route->group;
 
                 // Check if this Request is legit to be forwarded to upstream
-                for(sn = upstrIf->allowedgroups; sn != NULL; sn = sn->next)
+                for(sn = upstrIf->allowedgroups; sn != NULL; sn = sn->next) {
+                    // Check if there is a blacklist
+                    if (!sn->allow)
+                        block_list = true;
                     if((group & sn->subnet_mask) == sn->subnet_addr)
-                        // Forward is OK...
-                        break;
+                        match = sn;
+                }
 
-                if (sn == NULL) {
+                if((block_list && match != NULL && !match->allow) ||
+                  (!block_list && match == NULL)) {
                     my_log(LOG_INFO, 0, "The group address %s may not be forwarded upstream. Ignoring.", inetFmt(group, s1));
```

Comment:
This condition evaluated incorrectly for address `3` and following configuration:
```
whitelist 1
blacklist 2
```
Because `block_list` is set to true and `match` to `NULL`. Which results in forwarding address `3`. But based on logic in `request.c` and description in manpage it should not be forwarded.

Reply:
Damnit, I'm starting to regret this PR ;-)

Should be fixed now.

I copied the evaluation from request.c, and negate the result.

```
                // Keep in sync with request.c, note the negation
                if(!((!allow_list && match == NULL) ||
                  (allow_list && match != NULL && match->allow))) {
```

Reply:
Ok, this seems to be correct now.

And, well this is probably reason why nobody wanted to fix issue #54 for a longer time...